### ISSUE TITLE
Provide a public method in forcefield.py to get templates and list unmatched residues

### ISSUE THREAD

**Issue by tristanic** [openmm/openmm#2725]
It would be *really* useful if `ForceField` had a public method for assigning templates to residues that instead of raising a `ValueError` on the first failed match, instead returns something like a tuple containing `({matched_residue: template}, [unmatched_residues])`. The only currently supported approach is to try running `ForceField.createSystem`, catch the exception and parse the offending residue from the error text. For an interactive environment like ISOLDE that can get *really* slow - Python takes ages to unroll everything on each iteration, and for a model with many unmatched residues it becomes a painful cycle of try/assess/fix/try/assess/fix. Being able to get all unmatched residues at once would make for a *far* smoother user experience.

**Comment by tristanic** [comment]
Ahem. Now that I look at it... are `getMatchingTemplates()` and `getUnmatchedResidues()` new, or am I just blind?

**Comment by tristanic** [comment]
... although it does seem like it would be more efficient if these were rolled together into a single method to avoid looping over everything twice.

**Comment by peastman** [comment]
Also note that things get really complicated if the force field uses either template generators (you're unlikely to run into that case) or patches (CHARMM36 does).  For example, if you have a disulfide bond then neither of the participating residues will match any template on its own.  But it can make them match by applying a patch that adds the bond between them.  `getMatchingTemplates()` and `getUnmatchedResidues()` don't handle patches.

**Comment by tristanic** [comment]
Yeah - at the moment I'm only supporting AMBER within ISOLDE. Don't have
the resources to do more than that at this stage. I already have some
simple code in place to set the different cysteine variants (normal,
disulfide, deprotonated, bound to an iron-sulfur cluster), and to pull
common ligands into the forcefield from a zip file of individual ffXML
definitions. But I have more work to do to smoothly handle the
situations where there is no existing definition or the residue is
wrong/missing atoms.

On 2020-06-03 21:03, peastman wrote:

> Also note that things get really complicated if the force field uses either template generators (you're unlikely to run into that case) or patches (CHARMM36 does). For example, if you have a disulfide bond then neither of the participating residues will match any template on its own. But it can make them match by applying a patch that adds the bond between them. getMatchingTemplates() and getUnmatchedResidues() don't handle patches. 
> 
> --
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub [1], or unsubscribe [2].
 

Links:
------
[1] https://github.com/openmm/openmm/issues/2725#issuecomment-638431740
[2]

**Comment by jchodera** [comment]
> pull common ligands into the forcefield from a zip file of individual ffXML definitions

We've talked about providing a pre-parameterized set of ligands from the [PDB Ligand Expo](http://ligand-expo.rcsb.org/) in ffxml format for GAFF and openff. Would this be useful?

Structural ions are much more difficult. @rafwiewiora helped add some support for these, but it comes down to knowing which residues require protonation state modifications to coordinate the mutlisite ion models (which seem to be the most convenient).

For post-translational modified residues, we're hoping that the additions we're making to the `openforcefield` toolkit to parameterize AMBER-like biopolymers and small molecules consistently will be useful soon to provide a self-consistent way to handle any biopolymer and biomolecule you throw at it.

**Comment by tristanic** [comment]
@jchodera that would certainly be useful. The set I have at the moment (derived from https://github.com/nwmoriarty/amber_library) is incomplete and not particularly thoroughly validated - although most behave reasonably well, I have come across a few that are troublesome.

Regarding metals, they're difficult no matter which way you slice it. So much variability in their coordination states and geometry... ISOLDE provides a mechanism for user-adjustable harmonic distance restraints (add/remove/adjust distance or spring constant)... in the medium term, I'll probably just add some code to recognise at least the most common types of metal coordination and add these to suit, leaving it up to the user to fine-tune where they disagree with density. Longer term... well, for a start I'm hopeful that the polarisable force fields will do a better job with these, but I'm not sure they're mature enough yet to use in an application like this.

**Comment by tristanic** [comment]
I'd like to suggest something like the below as an improved alternative to `getMatchingTemplates`, to get unique matches, multiple matches and unmatched residues in a single iteration through the topology. This will make it much easier for interactive code like my own to guide the user through correcting problems with the model.

```
    def getMatchingTemplates(self, topology, ignoreExternalBonds=False):
        '''
        Returns three items:

            - a {residue: (template, atom_order)} dict for all residues in the topology for
              which a unique template was found
            - a {residue: [(template, atom_order)]} dict for any residues with multiple
              matching templates
            - a list of residues with no matching template.

        For a simulation to start the second and third objects should be empty.
        '''
        from simtk.openmm.app.forcefield import _createResidueSignature
        from simtk.openmm.app.internal import compiled
        bondedToAtom = self._buildBondedToAtomList(topology)
        unique_matches = {}
        multiple_matches = {}
        unmatched = []

        templateSignatures = self._templateSignatures
        for res in topology.residues():
            sig = _createResidueSignature([atom.element for atom in res.atoms()])
            if sig in templateSignatures:
                allMatches = []
                for t in templateSignatures[sig]:
                    match = compiled.matchResidueToTemplate(res, t, bondedToAtom, ignoreExternalBonds)
                    if match is not None:
                        allMatches.append((t, match))
                if len(allMatches) == 1:
                    unique_matches[res] = allMatches[0]
                elif len(allMatches) > 1:
                    multiple_matches[res] = allMatches
                else:
                    unmatched.append(res)
        return unique_matches, multiple_matches, unmatched
```

**Comment by jchodera** [comment]
@tristanic: Is there a way we can extend the API to give you what you need without breaking the API? We marked the API as experimental, but it would be less disruptive if the behavior could be modified via an optional argument or another method could be added.

**Comment by tristanic** [comment]
@jchodera: Fair point. For the time being I can implement this on my end by subclassing `ForceField`. Still tinkering around - for example, I realised it should also have the optional argument for a user provided {residue: template name} dict (for those entries it would just double-check that the template actually matches). Once I'm happy with that I'll get back to you. The only thing then would be to just choose a name that doesn't clash with existing methods (`prepareTemplates` perhaps?).

**Comment by tristanic** [comment]
A successful run of this method (that is, one where only `unique_matches` is populated) would provide all the information that's currently provided by `_matchAllResiduesToTemplates` during `createSystem`. So the next thing on the wish-list would be to be able to pass a pre-created `templateForResidue` to `createSystem` to avoid doubling up on work.
